AltinnRowId support

ivarne opened the pull request:
This PR assumes models will be extended with the following property

```C#
    [XmlAttribute("altinnRowId")]
    [JsonPropertyName("altinnRowId")]
    [System.Text.Json.Serialization.JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingDefault)]
    public Guid AltinnRowId { get; set; };

    public bool AltinnRowIdSpecified()
    {
        return AltinnRowId != default;
    }

```
## Description
PS: There should be more automated tests for this, but I don't think that prevents review.

Most significant changes:
* Added code for initialising and cleaning AltinnRowIds
* `IUserAction` related endpoints now runs validation
* fixed an error in `UserActionResult` that stated that dataType should be used as key for the updated models instead of dataElementId.
* Declare the result dictionary in `IUserAction` to be not nullable for the updated elements. It makes no sense to specify an element without the updated data for that element.

## Related Issue(s)
- https://github.com/Altinn/app-frontend-react/issues/1812

## Verification
- [ ] **Your** code builds clean without any errors or warnings
- [ ] Manual testing done (required)
- [ ] Relevant automated test added (if you find this hard, leave it and we'll help out)
- [ ] All tests run green

## Documentation
- [ ] User documentation is updated with a separate linked PR in [altinn-studio-docs.](https://github.com/Altinn/altinn-studio-docs) (if applicable)

ivarne committed:
First draft for AltinnRowId support

ivarne committed:
Fix support for IUserActions

Commit:
Mini fix to make saving new data models + validations to work

ivarne commented on src/Altinn.App.Api/Controllers/ActionsController.cs:
```diff
@@ -156,7 +156,7 @@ public class ActionsController : ControllerBase
     {
         var instanceIdentifier = new InstanceIdentifier(instance);
-        foreach (var (dataType, newModel) in resultUpdatedDataModels)
+        foreach (var (elementId, newModel) in resultUpdatedDataModels)
```
The comment about what the IUserAction should return states 

https://github.com/Altinn/app-lib-dotnet/blob/3f2dbd03700122cf9f59d3350686010f73513900/src/Altinn.App.Core/Models/UserAction/UserActionResult.cs#L17-L20

But I'm really happy if frontend used the more sensible elementId from the start. It removes one paragraph from https://github.com/Altinn/app-lib-dotnet/issues/446

olemartinorg replied:
Huh, I didn't know - that comment must be wrong. The key is, and must be, the elementId. I changed this here because existing user actions broke. :thinking:

ivarne replied:
If it broke existing actions that is really great 🎉🎉🎉🎉

Much easier to fix the comment than to make kompatibility with both options.

tjololo replied:
Yeah, sorry my bad! Comment and code is really misleading here, the key is elementId and always was so this is just a renaming luckily!

olemartinorg commented:
One more path that creates models but don't currently inject row IDs: `IInstantiationProcessor.DataCreation()`. We have a test for this in `ttd/frontend-test`, where we create Likert rows this way.

ivarne commented:
> One more path that creates models but don't currently inject row IDs: IInstantiationProcessor.DataCreation(). We have a test for this in ttd/frontend-test, where we create Likert rows this way.

What API gets called in that case? As far as I'm able to read all code paths leads to an api that returns a url or a `DataElement`, which as far as I understand means that frontend needs to use the GET endpoint to fetch the data element, witch initialises row ids.

ivarne committed:
Merge branch 'v8' into ivarne/altinnRowId

ivarne committed:
Fix build errors from validation change

Also make the return type from IUserAction require a model for all returned data elements.

olemartinorg commented:
> What API gets called in that case? As far as I'm able to read all code paths leads to an api that returns a url or a `DataElement`, which as far as I understand means that frontend needs to use the GET endpoint to fetch the data element, witch initialises row ids.

Strange, it's just called via the normal GET endpoint. No row IDs are added. You can test this yourself in the `ttd/frontend-test` app, with the `unique-row-ids` branch (probably best to run app-frontend locally with the `feat/row-uuids` branch as well). When starting the app, you can click the custom button to skip to a task, and then skip to the Likert task. It gives me this data model on the GET endpoint:

```json
{
  "Questions": [
    {
      "Id": "question-1",
      "Answer": ""
    },
    {
      "Id": "question-2",
      "Answer": ""
    },
    {
      "Id": "question-3",
      "Answer": ""
    },
    {
      "Id": "question-4",
      "Answer": ""
    },
    {
      "Id": "question-5",
      "Answer": ""
    },
    {
      "Id": "question-6",
      "Answer": ""
    }
  ]
}
```

ivarne commented:
Did you add the `AltinnRowId` guid, manually on the row element C# class, or did you run the update tool I have not written yet?

olemartinorg commented:
Yup, I added it manually (did not use the tool that doesn't exist! :sweat_smile:) It all works if I manually add row IDs in the `IInstantiationProcessor` where this data is first created, but I expect that not to be needed.

olemartinorg commented:
Btw, it also seems that we get row IDs back when fetching the data models after the instance has been submitted. We have automatic tests that compare those models to a ground truth, and they currently get the data models with row IDs. 

We discussed using a special header or query parameter when fetching the model from app-frontend so that the row IDs do not show up for other users. Let me know what you think, so I can add such a header/query parameter.

ivarne commented:
I think I prefer an optional GET parameter, but need some help on naming. `includeRowId`, `includeMetadata`, `includeInternalFields`, `frontendMetadata`?

olemartinorg commented:
I think I prefer `includeRowId` if we're going for the explicit `altinnRowId` field, otherwise probably `frontendMetadata` if we're settling on a metadata object.

ivarne committed:
Merge remote-tracking branch 'origin/v8' into ivarne/altinnRowId

ivarne committed:
Initialize AltinnRowId in more places

ivarne commented:
~~Æsj, parameteren ble hetende initializeAltinnRowId, for jeg klarte ikke finne ut hvor du hadde kommet med forslaget (trodde det var slack).~~ **Fixet nå**

Uansett har jeg nå lagt til initialisering av AltinnRowId en masse nye steder, men det var bare GET data der det ga mening å ha en parameter for å velge.

Håper å få fikset mer tester etter hvert.

ivarne committed:
Fix naming of parameter

ivarne committed:
Fix testst that broke

olemartinorg commented:
I ran our full e2e test-suite for `ttd/frontend-test` before the last commit (along with the changes in app-frontend I'm working on), and everything passed! :partying_face:

ivarne committed:
Add test and fix style issues

ivarne committed:
Fix AltinnRowId in tests to conform to current upgrade tool

RonnyB71 commented on src/Altinn.App.Api/Controllers/DataController.cs:
```diff
@@ -289,7 +291,7 @@ private static bool FileValidationEnabledForDataType(DataType dataTypeFromMetada
                 else if (dataType.AppLogic?.ClassRef is not null)
                 {
-                    return await GetFormData(org, app, instanceOwnerPartyId, instanceGuid, dataGuid, dataType, language, instance);
+                    return await GetFormData(org, app, instanceOwnerPartyId, instanceGuid, instance, dataGuid, dataElement, dataType, includeRowId, language);
```
Minor, but do we need both instance and instanceGuid? Since instance is introduced, perhaps just remove instanceGuid and get the id from instance object.

ivarne replied:
It is just that `Guid.Parse(instance.Id.Split("/")[1])` gets a little repetitive.

RonnyB71 commented on src/Altinn.App.Api/Controllers/InstancesController.cs:
```diff
@@ -844,6 +844,8 @@ private async Task CopyDataFromSourceInstance(ApplicationMetadata application, I
 
                     await _instantiationProcessor.DataCreation(targetInstance, data, null);
 
+                    ObjectUtils.InitializeAltinnRowId(data);
```
Candidate method for moving to a future DataService implementation instead of Utils class, but that should also include refactoring  this controller to not use the IDataClient directly but rather use the DataService and perhaps refactor even more logic out from the controller. But not critical now I think.

ivarne replied:
Yes, out of scope now, but would be great if we had test coverage to ensure that the future rewrite doesn't break the behavior.

RonnyB71 commented on src/Altinn.App.Core/Helpers/ObjectUtils.cs:
```diff
@@ -10,13 +10,25 @@ public static class ObjectUtils
         foreach (var prop in model.GetType().GetProperties())
         {
-            if (prop.PropertyType.IsGenericType && prop.PropertyType.GetGenericTypeDefinition() == typeof(List<>))
+            if (prop.PropertyType == typeof(Guid) && prop.Name == "AltinnRowId")
```
Perhaps a simple refactor of this to a `PropertyIsAltinnRowGuid(PropertyInfo prop)` just remove the duplicated logic in the `RemoveAltinnRowIds `method and keep the string name in one place.

```
    private static bool PropertyIsAltinnRowGuid(System.Reflection.PropertyInfo prop)
    {
        return prop.PropertyType == typeof(Guid) && prop.Name == "AltinnRowId";
    }
```

ivarne replied:
Done

RonnyB71 commented on src/Altinn.App.Core/Implementation/DefaultTaskEvents.cs:
```diff
@@ -160,20 +160,77 @@ public async Task OnEndProcessTask(string endEvent, Instance instance)
+        object data = await _dataClient.GetFormData(
+            instanceGuid, modelType, instance.Org, app, instanceOwnerPartyId, dataGuid);
+
+        // Remove hidden data before validation, ignore hidden rows. TODO: Determine how hidden rows should be handled going forward.
```
Is this TODO still relevant?

ivarne replied:
No, this one has been fixed.

ivarne replied:
```suggestion
        // Remove hidden data before validation, ignore hidden rows.
```

RonnyB71 commented on src/Altinn.App.Core/Implementation/DefaultTaskEvents.cs:
```diff
@@ -160,20 +160,77 @@ public async Task OnEndProcessTask(string endEvent, Instance instance)
+            }
         }
+        // remove AltinnRowIds
+        ObjectUtils.RemoveAltinnRowId(data);
```
Should we always remove the row id upon completion?

olemartinorg replied:
Yup, I believe so. They're strictly only needed for us, internally when filling out a form. If app developers have other needs - such as requiring that rows are unique in some way, where no rows may contain the same data as any other row - that is never handled by our row ID anyway, and would have to be implemented using validation messages.

EDIT: Also, importantly, as long as the row ID is not part of the XSD or JsonSchema, the data would not be valid when containing row IDs.

RonnyB71 reviewed (approved):
🕹️ approved we a few minor comments/questions.

ivarne committed:
Update src/Altinn.App.Core/Implementation/DefaultTaskEvents.cs

ivarne committed:
Extract method in ObjectUtils
